Upgraded gradle wrapper to 4.10.2 and fatjar to 0.3

RohanTalip opened the pull request:
### Checklist
- [x] I have made a material change to the repo (functionality, testing, spelling, grammar)
- [x] I have read the [Contribution Guide] and my PR follows them.
- [x] I updated my branch with the master branch.
- [ ] I have added tests that prove my fix is effective or that my feature works
  - N/A.
- [ ] I have added necessary documentation about the functionality in the appropriate .md file
  - N/A.
- [x] I have added in line documentation to the code I modified

### Short description of what this PR does:

Upgrading to a recent gradle version allows for a gradle composite build.

The gradle wrapper was upgraded by running this command (a few times while issues were being worked out) :
./gradlew wrapper --gradle-version=4.10 --distribution-type=bin

With fatjar still at 0.1.2, this error was showing up:

> Failed to apply plugin [id 'fatjar']
   > Could not find method add() for arguments [fatJarPrepareFiles, class eu.appsatori.gradle.fatjar.tasks.PrepareFiles] on task set of type org.gradle.api.internal.tasks.DefaultTaskContainer.

Upgrading to fatjar 0.3 fixed this issue.

This pull request is similar to pull request #436, with a newer version of gradle, but is slightly smaller and also doesn't change the build.gradle file away from using the unmaintained Gradle FatJar plugin to the Gradle Shadow plugin yet (but instead upgrades it from 0.1.2 to 0.3).

SendGridDX commented:
[![CLA assistant check](https://cla.sendgrid.com/pull/badge/not_signed)](https://cla.sendgrid.com/sendgrid/sendgrid-java?pullRequest=459) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.sendgrid.com/sendgrid/sendgrid-java?pullRequest=459) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.sendgrid.com/check/sendgrid/sendgrid-java?pullRequest=459) it.</sub>

thinkingserious commented:
Thanks @RohanTalip!

This has been added to our backlog for a code review.

With Best Regards,

Elmer

RohanTalip commented:
@thinkingserious, thanks but I see you said the same thing on https://github.com/sendgrid/sendgrid-java/pull/436 in May and that one still hasn't been reviewed from what I can tell.

Is this open source project still a priority for SendGrid?  I'm not trying to be snarky, just expressing genuine interest, as I don't see that many recent commits and there are quite a few old open pull requests that I would expect to either be merged or closed as infeasible or not meeting the direction of the product, within N days of being opened (where N is some reasonable number like 7 or 30).

I see that you added the difficulty:medium tag to this pull request.  If you or someone else at SendGrid would like to replicate what I've done, you can start with running "./gradlew wrapper --gradle-version=4.10 --distribution-type=bin" and then upgrading the fatjar dependency (or alternatively migrating to https://github.com/johnrengelman/shadow ).

I have no attachment to being the author for such a change (i.e. feel free to make this change yourselves), it would just be nice to have a recent version of Gradle embedded in the sendgrid-java project.

Regards,

Rohan

RohanTalip commented:
Actually, I see that Gradle 4.10.1 has been released now, so if you were going to make this change yourself, I would upgrade to that instead of 4.10: https://github.com/gradle/gradle/releases/tag/v4.10.1

thinkingserious commented:
Hi @RohanTalip, 

Your PR is actually coming up soon for a review. This Java SDK (just finished Python, Node.js, PHP and C#) is next on our list to get the dynamic template helpers added and we hope to merge in many of these open PRs that have been sitting for a very long time as part of that release.

If you would be so kind to update this PR with the updated 4.10.1 version or maybe 4.10.*, that would be awesome.

With Best Regards,

Elmer

RohanTalip commented:
Hi @thinkingserious , I had missed your update until now.  I just updated the gradle version from 4.10 to 4.10.2.

(Personally, if I was maintaining an open source project, I would be very cautious about accepting binary changes, such as .jar files, partly for the changes being opaque and partly for security issues such as local privilege escalation.)

thinkingserious commented:
That makes sense @RohanTalip, thanks!

thinkingserious commented:
HI @marshallpierce, 

Do you mind giving this a quick look in comparison to your PR?

Whoever's PR gets accepted in the end, both of you will be getting swag if you have not already :)

With Best Regards,

Elmer

marshallpierce commented:
They're pretty similar; I tackled the fatjar -> shadow conversion mentioned [here](https://github.com/sendgrid/sendgrid-java/pull/459/files#diff-c197962302397baf3a4cc36463dce5eaR36). As mentioned in the comments on mine, it's not clear what the original intent was for the fatjar, so maybe the conversion to use the shadow plugin needs additional config? It's pretty unusual for projects to generate a fat jar and put it in the project root :man_shrugging:

RohanTalip commented:
I see that @marshallpierce also updated his PR #436 to use 4.10.2.  I cloned his repository locally and switched to his build-on-recent-java branch:

$ mkdir marshallpierce
$ cd marshallpierce
$ git clone https://github.com/marshallpierce/sendgrid-java
$ cd sendgrid-java
$ git checkout build-on-recent-java

Comparing the 2 binary gradle-wrapper.jar files (my change was on the master branch of sendgrid-java, with my fork on GitHub as the default remote repository) :
$ cd ../..
$ shasum -a 256 sendgrid/sendgrid-java/gradle/wrapper/gradle-wrapper.jar marshallpierce/sendgrid-java/gradle/wrapper/gradle-wrapper.jar 
ad63ba21fb91e490e0f6fd0ca7d4049241f0f68a454b0b3075c041c4554e611c  sendgrid/sendgrid-java/gradle/wrapper/gradle-wrapper.jar
ad63ba21fb91e490e0f6fd0ca7d4049241f0f68a454b0b3075c041c4554e611c  marshallpierce/sendgrid-java/gradle/wrapper/gradle-wrapper.jar

Double checking:
$ diff -s sendgrid/sendgrid-java/gradle/wrapper/gradle-wrapper.jar marshallpierce/sendgrid-java/gradle/wrapper/gradle-wrapper.jar 
Files sendgrid/sendgrid-java/gradle/wrapper/gradle-wrapper.jar and marshallpierce/sendgrid-java/gradle/wrapper/gradle-wrapper.jar are identical

Checking the scripts:

$ shasum -a 256 sendgrid/sendgrid-java/gradlew* marshallpierce/sendgrid-java/gradlew*
8c4c04dd98db1f00d49456dd162418a39312c5cb13d6865d783deb483bd1ed22  sendgrid/sendgrid-java/gradlew
0008d785920c9ff5cab17403e0270ccc7ceee8e169b6d67a82d96a5475fec5c9  sendgrid/sendgrid-java/gradlew.bat
8c4c04dd98db1f00d49456dd162418a39312c5cb13d6865d783deb483bd1ed22  marshallpierce/sendgrid-java/gradlew
0008d785920c9ff5cab17403e0270ccc7ceee8e169b6d67a82d96a5475fec5c9  marshallpierce/sendgrid-java/gradlew.bat

$ diff -s sendgrid/sendgrid-java/gradlew marshallpierce/sendgrid-java/gradlew
Files sendgrid/sendgrid-java/gradlew and marshallpierce/sendgrid-java/gradlew are identical

$ diff -s sendgrid/sendgrid-java/gradlew.bat marshallpierce/sendgrid-java/gradlew.bat
Files sendgrid/sendgrid-java/gradlew.bat and marshallpierce/sendgrid-java/gradlew.bat are identical

The build.gradle file is obviously going to be different between our 2 PRs.

@marshallpierce or you @thinkingserious could reproduce these checks ...

thinkingserious commented:
Thanks for the responses @marshallpierce and @RohanTalip!

One of the goals is to produce a fat jar that folks can download and add to their project directly in case they don't want to use a package manager. In the end I move that fatJar out and publish for direct download.

marshallpierce commented:
That's fine, but you don't have to do the janky "host a jar on GitHub" system. You can publish a fat jar with different coordinates via bintray, etc...

thinkingserious commented:
Hi @marshallpierce,

I'm definitely open to that idea, could you please elaborate on the steps or update your PR to help us migrate to a more sensible solution? Thanks!

marshallpierce commented:
I don't have the time to implement the build changes at the moment, but here's the idea: currently, this library is published at the coordinates `com.sendgrid:sendgrid-java` with the usual jar, pom, etc. You can see all the stuff for, say, 4.2.1 [here](https://repo1.maven.org/maven2/com/sendgrid/sendgrid-java/4.2.1/). It's feasible to do the same thing, but with different coordinates (say, `com.sendgrid:sendgrid-java-nodeps`). So, one release would now generate 2 jars (one fat, one normal), and all the associated pom, etc, for each one. 

That said, instead of tinkering with that, you might be better off helping customers use a build tool, like with a sample project that has Gradle set up already and is designed for customers to use as a starting point, or something like that... Projects without build automation are sad projects.

thinkingserious commented:
That's awesome feedback @marshallpierce,

Would you open to helping craft a new issue to that end?

thinkingserious commented:
Hello @RohanTalip,<br /><br/>Thanks again for the PR!<br />                                              <br />It's <a href="https://sendgrid.com/blog/hacktoberfest-2018-has-arrived/">HACKTOBERFEST</a>! We want to show our appreciation by sending                                              you some special Hacktoberfest swag. If you have not already, could you please fill out                                              <a href="https://dx.sendgrid.com/hacktoberfest/join">this form</a> so we can send it to you? Thanks!                                             <br /><br />Team SendGrid DX

thinkingserious commented:
Hello @RohanTalip,<br /><br/>Thanks again for the PR!<br /><br />                                         We noticed that you have earned 3 out of the 5 points needed to receive glorious SendGrid Hacktoberfest swag.<br /><br />                                         Please take a moment to checkout <a href="https://dx.sendgrid.com/hacktoberfest/issues">this link</a> to find more issues to get you past the required threshold.<br /><br />                                         Also, please be sure that you have officially registered <a href="https://dx.sendgrid.com/hacktoberfest/join">here</a>.<br /><br />                                         Thank you and Happy Hacktobering!<br /><br />                                         Team SendGrid DX

thinkingserious commented:
Hello @RohanTalip,<br /><br/>Thanks again for the PR!<br /><br />                                         You have earned 3 out of the 5 points needed to receive glorious SendGrid Hacktoberfest swag.<br /><br />                                         Please take a moment to checkout <a href="https://dx.sendgrid.com/hacktoberfest/issues">this link</a> to find more issues to get you past the required threshold or to simply continue the celebration.<br /><br />                                         Also, <strong>please be sure you have officially registered with us <a href="https://dx.sendgrid.com/hacktoberfest/join">here</a> by November 1, 2018</strong> to qualify.<br /><br />                                         If you have any questions you can email us at <a href="mailto:dx+hacktoberfest2018@sendgrid.com">dx+hacktoberfest2018@sendgrid.com</a>.<br /><br />                                         Thank you and Happy Hacktobering!<br /><br />                                         Team SendGrid

RohanTalip commented:
This pull request is really old now and it looks like this project is now going to be using Maven rather than Gradle, so I'm going to close this PR.
